haystack/generator/transformers.py

```diff
@@ -18,91 +19,227 @@ class RAGGeneratorType(Enum):
+        self.passage_encoder = DPRContextEncoder.from_pretrained(passage_embedding_model).to(self.device)
+
+    # Copied cat_input_and_doc method from transformers.RagRetriever
+    def _cat_input_and_doc(self, doc_title: str, doc_text: str, input_string: str, prefix: Optional[str]):
+        if doc_title.startswith('"'):
+            doc_title = doc_title[1:]
+        if doc_title.endswith('"'):
+            doc_title = doc_title[:-1]
+        if prefix is None:
+            prefix = ""
+        out = (prefix + doc_title + self.model.config.title_sep + doc_text + self.model.config.doc_sep +
+               input_string).replace("  ", " ")
+
+        return out
 
-    # TODO: Copy from RagRetriever and modified
-    # Need to understand the logic
-    def _postprocess_docs(self, retrieved_texts: List[str], question: str, return_tensors: str = "pt"):
-        def cat_input_and_doc(doc_text, input_string):
-            out = (" / " + doc_text + " // " + input_string).replace("  ", " ")
-            return out
+    # Copied postprocess_docs method from transformers.RagRetriever and modified
+    def _get_contextualized_embedding(self, texts: List[str], question: str, titles: Optional[List[str]] = None,
+                                      return_tensors: str = "pt"):
+
+        titles_list = titles if self.embed_title and titles is not None else [""] * len(texts)
+        prefix = self.prefix if self.prefix is not None else self.model.config.generator.prefix
 
         rag_input_strings = [
-            cat_input_and_doc(
-                retrieved_texts[i],
-                question,
+            self._cat_input_and_doc(
+                doc_title=titles_list[i],
+                doc_text=texts[i],
+                input_string=question,
+                prefix=prefix
             )
-            for i in range(len(retrieved_texts))
+            for i in range(len(texts))
         ]
 
-        contextualized_inputs = self.tokenizer.question_encoder.batch_encode_plus(
+        contextualized_embedding = self.tokenizer.question_encoder.batch_encode_plus(
             rag_input_strings,
-            max_length=300,
+            max_length=self.model.config.max_length,
             return_tensors=return_tensors,
             padding="max_length",
             truncation=True,
         )
 
-        return contextualized_inputs["input_ids"], contextualized_inputs["attention_mask"]
+        return contextualized_embedding["input_ids"].to(self.device), \
+               contextualized_embedding["attention_mask"].to(self.device)
+
+    def _embed_passage(self, title: str, text: str) -> torch.Tensor:
+        """Compute the DPR embeddings of document passages"""
+        input_ids = self.passage_tokenizer(
+            title,
+            text,
+            truncation=True,
+            padding="longest",
+            return_tensors="pt"
+        )["input_ids"]
+
+        embeddings = self.passage_encoder(
```

Comment:
We can speed this up significantly by not passing individual passages here but batches of passages. Every forward pass in the model is quite costly.
I would suggest getting rid of `_embed_passage()` and refactor `embed_passage_in_tensor()` to be very similar to DPR's `embed_passages()` https://github.com/deepset-ai/haystack/blob/master/haystack/retriever/dense.py#L103

This will also simplify maintenance when we have just one way of creating DPR passage embeddings.

Reply:
Okay now passing DensePassageRetriever so directly embed_passage function can be utilized